apps/rush-lib/src/logic/taskRunner/AsyncTaskQueue.ts

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { Task } from './Task';
+import { TaskStatus } from './TaskStatus';
+
+/**
+ * @internal
+ *
+ * Implmentation of the async iteration protocol for a collection of Task objects.
+ * The async iterator will wait for a task to be ready for execution, or terminate if there are no more tasks.
+ *
+ * @remarks
+ * It is imperative that the caller update the status of tasks prior to invoking `next()` on the iterator again,
+ * otherwise it will stall forever unless some other call stack invokes next().
+ */
+export class AsyncTaskQueue implements AsyncIterable<Task>, AsyncIterator<Task> {
```

Comment:
This is neat. Is this something that could replace the task logic used to run project actions in Rush? Like running `rush build` would create a new async task for each project and add to the queue?

Reply:
That's exactly how `rush build` works internally. This class just takes the dependency graph of tasks and iterates them in topological order as they are completed, but part of the contract (deleting dependencies that have been satisfied) is still handled by the caller.